File: naive_bayes.r
############### problem1 ######################
#1.) Prepare a classification model using Naive Bayes for Salary dataset, train and test datasets are given separately use both datasets for model building. 	

# Import  dataset
library(readr)
salary_test <- read.csv(file.choose())
salary_train <- read.csv(file.choose())

#Data preprocessing for test data
salary_test$Salary <- ifelse(salary_test$Salary == " <=50K",0,1)

str(salary_test)
colnames(salary_test)

#factor analysis for test data
cols <- c("age","workclass","education","educationno","maritalstatus","occupation","relationship","race","sex","capitalgain","capitalloss","hoursperweek","native","Salary")
salary_test[cols] <- lapply(salary_test[cols],factor)
str(salary_test)


#Data preprocessing for train data
salary_train$Salary <- ifelse(salary_train$Salary == " <=50K",0,1)

str(salary_train)
colnames(salary_train)

#factor analysis for train data
col <- c("age","workclass","education","educationno","maritalstatus","occupation","relationship","race","sex","capitalgain","capitalloss","hoursperweek","native","Salary")
salary_train[col] <- lapply(salary_train[col],factor)
str(salary_train)


#viewing str and table
str(salary_test$Salary)
table(salary_test$Salary)

str(salary_train$Salary)
table(salary_train$Salary)

# proportion of salaries 0(<=50k) and 1(>50k)
prop.table(table(salary_test$Salary))
prop.table(table(salary_train$Salary))

##  Training a model on the data ----
install.packages("e1071")
library(e1071)
## building naiveBayes classifier.
salary_classifier <- naiveBayes(Salary~., data = salary_train)
salary_classifier

### laplace smoothing, by default the laplace value = 0
## naiveBayes function has laplace parameter, the bigger the laplace smoothing value, 
# the models become same.
salary_lap <- naiveBayes(Salary~., data = salary_train,laplace = 3)
salary_lap

##  Evaluating model performance without laplace
salary_test_pred <- predict(salary_classifier, salary_test)

# Evaluating model performance after applying laplace smoothing
salary_test_pred_lap <- predict(salary_lap, salary_test)

## crosstable without laplace
install.packages("gmodels")
library(gmodels)

CrossTable(salary_test_pred, salary_test$Salary,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))
## test accuracy
test_acc <- mean(salary_test_pred == salary_test$Salary)
test_acc

## crosstable of laplace smoothing model
CrossTable(salary_test_pred_lap, salary_test$Salary,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))

## test accuracy after laplace 
test_acc_lap <- mean(salary_test_pred_lap == salary_test$Salary)
test_acc_lap

# On Training Data without laplace 
salary_train_pred <- predict(salary_classifier, salary_train)
salary_train_pred

# train accuracy
train_acc = mean(salary_train_pred == salary_train$Salary)
train_acc

# prediction on train data for laplace model
salary_train_pred_lap <- predict(salary_lap,salary_train)
salary_train_pred_lap

# train accuracy after laplace
train_acc_lap = mean(salary_train_pred_lap == salary_train$Salary)
train_acc_lap

####################### problem2 ##################################
#Problem Statement: -
#This dataset contains information of users in social network. This social network has several business clients which can put their ads on social network and one of the Client has a car company who has just launched a luxury SUV for ridiculous price. Build the Bernoulli Naïve Bayes model using this dataset and classify which of the users of the social network are going to purchase this luxury SUV.
#Purchased: - 1 and Not Purchased: - 0

# Import the dataset
library(readr)
suv_car <- read.csv(file.choose())
suv_car <- suv_car[c(-1)]
View(suv_car)
str(suv_car)
colnames(suv_car)

#data preprocessing
col <- c("Gender","Age","EstimatedSalary","Purchased")
col <- c("Purchased")
suv_car[col] <- lapply(suv_car[col],factor)
suv_car[c(3)] <- scale(suv_car[c(3)])
str(suv_car)

# examine the type variable more carefully
str(suv_car$Purchased)
table(suv_car$Purchased)

# proportion of 0(not purchased) and 1(purchased) messages
prop.table(table(suv_car$Purchased))

str(suv_car)

# creating training and test datasets
suv_car_train <- suv_car[1:280, ]
suv_car_test  <- suv_car[280:400, ]


# check that the proportion of 0(not purchased) and 1(purchased)  is similar
prop.table(table(suv_car$Purchased))

prop.table(table(suv_car_train$Purchased))
prop.table(table(suv_car_test$Purchased))

##  Training a model on the data ----
install.packages("e1071")
library(e1071)
## building naiveBayes classifier.
suv_car_classifier <- naiveBayes(Purchased~., data = suv_car_train)
suv_car_classifier

?naiveBayes
### laplace smoothing, by default the laplace value = 0
## naiveBayes function has laplace parameter, the bigger the laplace smoothing value, 
# the models become same.
suv_car_lap <- naiveBayes(Purchased~.,data = suv_car_train ,laplace = 3)
suv_car_lap

##  Evaluating model performance with out laplace
suv_car_test_pred <- predict(suv_car_classifier, suv_car_test)

# Evaluating model performance after applying laplace smoothing
suv_car_test_pred_lap <- predict(suv_car_lap, suv_car_test)

## crosstable without laplace
install.packages("gmodels")
library(gmodels)

CrossTable(suv_car_test_pred, suv_car_test$Purchased,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))

## test accuracy
test_acc <- mean(suv_car_test_pred ==suv_car_test$Purchased)
test_acc

## crosstable of laplace smoothing model
CrossTable(suv_car_test_pred_lap, suv_car_test$Purchased,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))

## test accuracy after laplace 
test_acc_lap <- mean(suv_car_test_pred_lap == suv_car_test$Purchased)
test_acc_lap

# On Training Data without laplace 
suv_car_train_pred <- predict(suv_car_classifier, suv_car_train)
suv_car_train_pred


# train accuracy
train_acc = mean(suv_car_train_pred == suv_car_train$Purchased)
train_acc


# prediction on train data for laplace model
suv_car_train_pred_lap <- predict(suv_car_lap,suv_car_train)
suv_car_train_pred_lap

# train accuracy after laplace
train_acc_lap = mean(suv_car_train_pred_lap == suv_car_train$Purchased)
train_acc_lap

################### problem3 ####################################
#Problem Statement: -
#In this case study you have been given with tweeter data collected from an anonymous twitter handle, with the help of Naïve Bayes algorithm predict a given tweet is Fake or Real about real disaster occurring. 
#Real tweet: - 1 and Fake tweet: - 0

# Import the twitter dataset
library(readr)
tweet <- read.csv(file.choose())
View(sms_raw)

str(tweet)

tweet$target <- factor(tweet$target)
# examine the type variable more carefully
str(tweet$target)
table(tweet$target)

# proportion of ham and spam messages
prop.table(table(tweet$target))

# build a corpus using the text mining (tm) package
install.packages("tm")
library(tm)
str(tweet$text)
tweet_corpus <- Corpus(VectorSource(tweet$text))
# str(sms_corpus)

tweet_corpus <- tm_map(tweet_corpus, function(x) iconv(enc2utf8(x), sub='byte'))
# clean up the corpus using tm_map()
corpus_clean <- tm_map(tweet_corpus, tolower)
corpus_clean <- tm_map(corpus_clean, removePunctuation)
corpus_clean <- tm_map(corpus_clean, removeNumbers)
corpus_clean <- tm_map(corpus_clean, removeWords, stopwords())
corpus_clean <- tm_map(corpus_clean, stripWhitespace)

# create a document-term sparse matrix
tweet_dtm <- DocumentTermMatrix(corpus_clean)
tweet_dtm
View(tweet_dtm[1:10, 1:30])
# To view DTM we need to convert it into matrix first
dtm_matrix <- as.matrix(tweet_dtm)
str(dtm_matrix)

View(dtm_matrix[1:10, 1:20])

colnames(tweet_dtm)[1:50]
# creating training and test datasets
tweets_train <- tweet[1:5329, ]
tweets_test  <- tweet[5330:7613, ]

tweet_corpus_train <- corpus_clean[1:5329]
tweet_corpus_test  <- corpus_clean[5330:7613]

tweet_dtm_train <-tweet_dtm[1:5329, ]
tweet_dtm_test  <- tweet_dtm[5330:7613, ]

# check that the proportion of spam is similar
prop.table(table(tweet$target)) 

prop.table(table(tweet_train$target))
prop.table(table(tweet_test$target))

# indicator features for frequent words
# dictionary of words which are used more than 5 times
tweet_dict <- findFreqTerms(tweet_dtm_train, 5)
tweet_train <- DocumentTermMatrix(tweet_corpus_train, list(dictionary = tweet_dict))
tweet_test  <- DocumentTermMatrix(tweet_corpus_test, list(dictionary = tweet_dict))

tweet_test_matrix <- as.matrix(tweet_test)
View(tweet_test_matrix[1:10,1:10])

# convert counts to a factor
# custom function: if a word is used more than 0 times then mention 1 else mention 0
convert_counts <- function(x) {
  x <- ifelse(x > 0, 1, 0)
  x <- factor(x, levels = c(0, 1), labels = c("No", "Yes"))
}

# apply() convert_counts() to columns of train/test data
# Margin = 2 is for columns
# Margin = 1 is for rows
tweet_train <- apply(tweet_train, MARGIN = 2, convert_counts)
tweet_test  <- apply(tweet_test, MARGIN = 2, convert_counts)
?apply()

View(tweet_test[1:10,1:10])

##  Training a model on the data ----
install.packages("e1071")
library(e1071)
## building naiveBayes classifier.
tweet_classifier <- naiveBayes(tweet_train, tweets_train$target)
tweet_classifier

### laplace smoothing, by default the laplace value = 0
## naiveBayes function has laplace parameter, the bigger the laplace smoothing value, 
# the models become same.
tweet_lap <- naiveBayes(tweet_train, tweets_train$target,laplace = 3)
tweet_lap

##  Evaluating model performance with out laplace
tweet_test_pred <- predict(tweet_classifier, tweet_test)

# Evaluating model performance after applying laplace smoothing
tweet_test_pred_lap <- predict(tweet_lap, tweet_test)

## crosstable without laplace
install.packages("gmodels")
library(gmodels)

CrossTable(tweet_test_pred, tweets_test$target,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))
## test accuracy
test_acc <- mean(tweet_test_pred == tweets_test$target)
test_acc

## crosstable of laplace smoothing model
CrossTable(tweet_test_pred_lap, tweets_test$target,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))

## test accuracy after laplace 
test_acc_lap <- mean(tweet_test_pred_lap == tweets_test$target)
test_acc_lap

# On Training Data without laplace 
tweet_train_pred <- predict(tweet_classifier, tweet_train)
tweet_train_pred

# train accuracy
train_acc = mean(tweet_train_pred == tweets_train$target)
train_acc


# prediction on train data for laplace model
tweet_train_pred_lap <- predict(tweet_lap,tweet_train)
tweet_train_pred_lap

# train accuracy after laplace
train_acc_lap = mean(tweet_train_pred_lap == tweets_train$target)
train_acc_lap

File: naive_bayes.py
##################### problem1 ###########################
#1.) Prepare a classification model using Naive Bayes for Salary dataset, train and test datasets are given separately use both datasets for model building. 	

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer,TfidfTransformer

# Loading the data set
salary_test = pd.read_csv("C:/Users/DELL/Downloads/SalaryData_Test.csv",encoding = "ISO-8859-1")
salary_train = pd.read_csv("C:/Users/DELL/Downloads/SalaryData_Train.csv",encoding = "ISO-8859-1")

# Preparing a naive bayes model on training data set 

from sklearn.naive_bayes import MultinomialNB as MB
x = {" <=50K" :1," >50K" :2}
salary_test.Salary = [x[item] for item in salary_test.Salary]
salary_train.Salary = [x[item] for item in salary_train.Salary]

#dummy variables for test data
dummies_salary_test = pd.get_dummies(salary_test)
dummies_salary_test.drop(['Salary'],axis = 1,inplace =True)
dummies_salary_test.isna()

#dummy variables for train data
dummies_salary_train = pd.get_dummies(salary_train)
dummies_salary_train.drop(['Salary'],axis = 1,inplace =True)
dummies_salary_train.isna()

# Multinomial Naive Bayes
classifier_mb = MB()
classifier_mb.fit(dummies_salary_train, salary_train.Salary)

# Evaluation on Test Data
test_pred_m = classifier_mb.predict(dummies_salary_test)
accuracy_test_m = np.mean(test_pred_m == salary_test.Salary)
accuracy_test_m

from sklearn.metrics import accuracy_score
accuracy_score(test_pred_m, salary_test.Salary) 

pd.crosstab(test_pred_m, salary_test.Salary)

# Training Data accuracy
train_pred_m = classifier_mb.predict(dummies_salary_train)
accuracy_train_m = np.mean(train_pred_m == salary_train.Salary)
accuracy_train_m

# Multinomial Naive Bayes changing default alpha for laplace smoothing
# if alpha = 0 then no smoothing is applied and the default alpha parameter is 1
# the smoothing process mainly solves the emergence of zero probability problem in the dataset.

classifier_mb_lap = MB(alpha = 3)
classifier_mb_lap.fit(dummies_salary_train, salary_train.Salary)

# Evaluation on Test Data after applying laplace
test_pred_lap = classifier_mb_lap.predict(dummies_salary_test)
accuracy_test_lap = np.mean(test_pred_lap == salary_test.Salary)
accuracy_test_lap

from sklearn.metrics import accuracy_score
accuracy_score(test_pred_lap, salary_test.Salary) 

pd.crosstab(test_pred_lap, salary_test.Salary)

# Training Data accuracy
train_pred_lap = classifier_mb_lap.predict(dummies_salary_train)
accuracy_train_lap = np.mean(train_pred_lap == salary_train.Salary)
accuracy_train_lap


######################## problem2 #####################
#Problem Statement: -
#This dataset contains information of users in social network. This social network has several business clients which can put their ads on social network and one of the Client has a car company who has just launched a luxury SUV for ridiculous price. Build the Bernoulli Naïve Bayes model using this dataset and classify which of the users of the social network are going to purchase this luxury SUV.
#Purchased: - 1 and Not Purchased: - 0

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer,TfidfTransformer

# Loading the data set
suv_car = pd.read_csv("C:/Users/DELL/Downloads/NB_Car_Ad.csv",encoding = "ISO-8859-1")

#dropping first column for analysis
suv_car.drop(['User ID'],axis = 1 ,inplace = True)

#scaling
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
suv_car [['Age','EstimatedSalary']] = scaler.fit_transform(suv_car [['Age','EstimatedSalary']])


# splitting data into train and test data sets 
from sklearn.model_selection import train_test_split

suv_car_train, suv_car_test = train_test_split(suv_car, test_size = 0.2)

# Preparing a naive bayes model on training data set 

from sklearn.naive_bayes import MultinomialNB as MB

#dummies for test and train data
suv_car_dumm_train = pd.get_dummies(suv_car_train)
suv_car_dumm_train.drop(["Purchased"], axis = 1, inplace = True)

suv_car_dumm_test = pd.get_dummies(suv_car_test)
suv_car_dumm_test.drop(["Purchased"], axis = 1, inplace = True)

# Multinomial Naive Bayes
classifier_mb = MB()
classifier_mb.fit(suv_car_dumm_train, suv_car_train.Purchased)

# Evaluation on Test Data
test_pred_m = classifier_mb.predict(suv_car_dumm_test)
accuracy_test_m = np.mean(test_pred_m == suv_car_test.Purchased)
accuracy_test_m

from sklearn.metrics import accuracy_score
accuracy_score(test_pred_m, suv_car_test.Purchased) 

pd.crosstab(test_pred_m, suv_car_test.Purchased)

# Training Data accuracy
train_pred_m = classifier_mb.predict(suv_car_dumm_train)
accuracy_train_m = np.mean(train_pred_m == suv_car_train.Purchased)
accuracy_train_m

# Multinomial Naive Bayes changing default alpha for laplace smoothing
# if alpha = 0 then no smoothing is applied and the default alpha parameter is 1
# the smoothing process mainly solves the emergence of zero probability problem in the dataset.

classifier_mb_lap = MB(alpha = 3)
classifier_mb_lap.fit(suv_car_dumm_train, suv_car_train.Purchased)


# Evaluation on Test Data after applying laplace
test_pred_lap = classifier_mb_lap.predict(suv_car_dumm_test)
accuracy_test_lap = np.mean(test_pred_lap == suv_car_test.Purchased)
accuracy_test_lap

from sklearn.metrics import accuracy_score
accuracy_score(test_pred_lap, suv_car_test.Purchased) 

pd.crosstab(test_pred_lap, suv_car_test.Purchased)

# Training Data accuracy
train_pred_lap = classifier_mb_lap.predict(suv_car_dumm_train)
accuracy_train_lap = np.mean(train_pred_lap == suv_car_train.Purchased)
accuracy_train_lap

################### problem3 #################################
#Problem Statement: -
#In this case study you have been given with tweeter data collected from an anonymous twitter handle, with the help of Naïve Bayes algorithm predict a given tweet is Fake or Real about real disaster occurring. 
#Real tweet: - 1 and Fake tweet: - 0

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer,TfidfTransformer

# Loading the data set
tweet = pd.read_csv("C:/Users/DELL/Downloads/Disaster_tweets_NB.csv",encoding = "ISO-8859-1")
tweet = tweet.iloc[:,3:5]
# cleaning data 
import re
stop_words = []
# Load the custom built Stopwords
with open("C:/Users/DELL/Downloads/stopwords_en.txt","r") as sw:
    stop_words = sw.read()

stop_words = stop_words.split("\n")
   
def cleaning_text(i):
    i = re.sub("[^A-Za-z" "]+"," ",i).lower()
    i = re.sub("[0-9" "]+"," ",i)
    w = []
    for word in i.split(" "):
        if len(word)>3:
            w.append(word)
    return (" ".join(w))

# testing above function with sample text => removes punctuations, numbers
tweet.text = tweet.text.apply(cleaning_text)

# removing empty rows
tweet = tweet.loc[tweet.text != " ",:]

# CountVectorizer
# Convert a collection of text documents to a matrix of token counts

# splitting data into train and test data sets 
from sklearn.model_selection import train_test_split

tweet_train, tweet_test = train_test_split(tweet, test_size = 0.2)

# creating a matrix of token counts for the entire text document 
def split_into_words(i):
    return [word for word in i.split(" ")]

# Defining the preparation of email texts into word count matrix format - Bag of Words
tweet_bow = CountVectorizer(analyzer = split_into_words).fit(tweet.text)

# Defining BOW for all messages
all_tweet_matrix = tweet_bow.transform(tweet.text)

# For training messages
train_tweet_matrix = tweet_bow.transform(tweet_train.text)

# For testing messages
test_tweet_matrix = tweet_bow.transform(tweet_test.text)

# Learning Term weighting and normalizing on entire emails
tfidf_transformer = TfidfTransformer().fit(all_tweet_matrix)

# Preparing TFIDF for train emails
train_tfidf = tfidf_transformer.transform(train_tweet_matrix)
train_tfidf.shape # (row, column)

# Preparing TFIDF for test emails
test_tfidf = tfidf_transformer.transform(test_tweet_matrix)
test_tfidf.shape #  (row, column)

# Preparing a naive bayes model on training data set 

from sklearn.naive_bayes import MultinomialNB as MB

# Multinomial Naive Bayes
classifier_mb = MB()
classifier_mb.fit(train_tfidf, tweet_train.target)

# Evaluation on Test Data
test_pred_m = classifier_mb.predict(test_tfidf)
accuracy_test_m = np.mean(test_pred_m == tweet_test.target)
accuracy_test_m

from sklearn.metrics import accuracy_score
accuracy_score(test_pred_m, tweet_test.target) 

pd.crosstab(test_pred_m, tweet_test.target)

# Training Data accuracy
train_pred_m = classifier_mb.predict(train_tfidf)
accuracy_train_m = np.mean(train_pred_m == tweet_train.target)
accuracy_train_m

# Multinomial Naive Bayes changing default alpha for laplace smoothing
# if alpha = 0 then no smoothing is applied and the default alpha parameter is 1
# the smoothing process mainly solves the emergence of zero probability problem in the dataset.

classifier_mb_lap = MB(alpha = 3)
classifier_mb_lap.fit(train_tfidf, tweet_train.target)

# Evaluation on Test Data after applying laplace
test_pred_lap = classifier_mb_lap.predict(test_tfidf)
accuracy_test_lap = np.mean(test_pred_lap == tweet_test.target)
accuracy_test_lap

from sklearn.metrics import accuracy_score
accuracy_score(test_pred_lap, tweet_test.target) 

pd.crosstab(test_pred_lap, tweet_test.target)

# Training Data accuracy
train_pred_lap = classifier_mb_lap.predict(train_tfidf)
accuracy_train_lap = np.mean(train_pred_lap == tweet_train.target)
accuracy_train_lap

################# END ########################################









